### PR TITLE
feat: implement `derived-has-same-inputs-outputs`

### PR DESCRIPTION
**Pull request by baseballyama**
Related to: https://github.com/ota-meshi/eslint-plugin-svelte/issues/198 (derived-inputs-outputs)

### CODE REVIEW

**Comment by baseballyama** [comment]
Maybe CI will be fixed after merging https://github.com/ota-meshi/eslint-plugin-svelte/pull/248 and https://github.com/ota-meshi/eslint-plugin-svelte/pull/247

File: src/rules/derived-has-same-inputs-outputs.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
+import type * as ESTree from "estree"
+import { createRule } from "../utils"
+import type { RuleContext } from "../types"
+import { extractStoreReferences } from "./reference-helpers/svelte-store"
+
+export default createRule("derived-has-same-inputs-outputs", {
+  meta: {
+    docs: {
+      description: "",
+      category: "Best Practices",
```
Comment (ota-meshi) [inline src/rules/derived-has-same-inputs-outputs.ts:10]: Perhaps `"Stylistic Issues"` is a more appropriate category.

```suggestion
      category: "Stylistic Issues",
```

File: src/rules/derived-has-same-inputs-outputs.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      const fnParam = fn.params[0]
+      if (fnParam.type !== "Identifier") return
+      const expectedName = `$${args.name}`
+      if (expectedName !== fnParam.name) {
+        context.report({
+          node: fn,
+          loc: {
+            start: fnParam.loc?.start ?? { line: 1, column: 0 },
+            end: fnParam.loc?.end ?? { line: 1, column: 0 },
+          },
```
Comment (ota-meshi) [inline src/rules/derived-has-same-inputs-outputs.ts:58]: Nodes handled by ESLint always have a loc property, so non-null expressions can be used.
https://eslint.org/docs/latest/developer-guide/working-with-custom-parsers#all-nodes

```suggestion
          loc: fnParam.loc!,
```

File: src/rules/derived-has-same-inputs-outputs.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      fn: ArrowFunctionExpressionOrFunctionExpression,
+    ) {
+      const fnParam = fn.params[0]
+      if (fnParam.type !== "ArrayPattern") return
+      const argNames = args.elements.map((element) => {
+        return element && element.type === "Identifier" ? element.name : null
+      })
+      fnParam.elements.forEach((element, index) => {
+        if (element && element.type === "Identifier") {
+          const expectedName = `$${argNames[index]}`
```
Comment (ota-meshi) [inline src/rules/derived-has-same-inputs-outputs.ts:81]: I think maybe if `argNames[index]` is `null` it should be ignored from the check.

File: src/rules/derived-has-same-inputs-outputs.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+          if (expectedName !== element.name) {
+            context.report({
+              node: fn,
+              loc: {
+                start: element.loc?.start ?? { line: 1, column: 0 },
+                end: element.loc?.end ?? { line: 1, column: 0 },
+              },
```
Comment (ota-meshi) [inline src/rules/derived-has-same-inputs-outputs.ts:88]:
```suggestion
              loc: element.loc!,
```

**Review by ota-meshi — CHANGES_REQUESTED** [review]
Thank you for this PR! I wrote some comments.

File: src/rules/derived-has-same-inputs-outputs.ts (hunk 4 above)
Comment (baseballyama) [reply src/rules/derived-has-same-inputs-outputs.ts]: fixed! https://github.com/ota-meshi/eslint-plugin-svelte/pull/249/commits/ba6dcccb7f8c182895f8084b2f951c4087aeaaa1
And my previous PR also should do the same way, so I updated.

File: src/rules/derived-has-same-inputs-outputs.ts (hunk 3 above)
Comment (baseballyama) [reply src/rules/derived-has-same-inputs-outputs.ts]: Sorry for that! I updated! https://github.com/ota-meshi/eslint-plugin-svelte/pull/249/commits/ba6dcccb7f8c182895f8084b2f951c4087aeaaa1

File: src/rules/derived-has-same-inputs-outputs.ts (hunk 2 above)
Comment (baseballyama) [reply src/rules/derived-has-same-inputs-outputs.ts]: fixed! https://github.com/ota-meshi/eslint-plugin-svelte/commit/ba6dcccb7f8c182895f8084b2f951c4087aeaaa1
And my previous PR also should do the same way, so I updated it.

File: src/rules/derived-has-same-inputs-outputs.ts (hunk 1 above)
Comment (baseballyama) [reply src/rules/derived-has-same-inputs-outputs.ts]: updated! https://github.com/ota-meshi/eslint-plugin-svelte/commit/ba6dcccb7f8c182895f8084b2f951c4087aeaaa1
`conflictWithPrettier` should be `false` right?🤔

File: src/rules/derived-has-same-inputs-outputs.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
+import type * as ESTree from "estree"
+import { createRule } from "../utils"
+import type { RuleContext } from "../types"
+import { extractStoreReferences } from "./reference-helpers/svelte-store"
+
+export default createRule("derived-has-same-inputs-outputs", {
+  meta: {
+    docs: {
+      description: "",
```
[the remark this answers is not in the document]
Comment (baseballyama) [reply src/rules/derived-has-same-inputs-outputs.ts]: ah sorry!
https://github.com/ota-meshi/eslint-plugin-svelte/commit/ba6dcccb7f8c182895f8084b2f951c4087aeaaa1

**Comment by baseballyama** [comment]
@ota-meshi And I added a few more tests. Thank you for your review!

**Review by ota-meshi — APPROVED** [review]
LGTM! Thank you for the great work!

**Review by JounQin — APPROVED** [review]

**Pull request merged**